src/lamp_py/bus_performance_manager/gtfs_rt_ingestion.py

```diff
@@ -0,0 +1,109 @@
+            )
+            & pl.col("vehicle.current_status").is_not_null()
+            & pl.col("vehicle.stop_id").is_not_null()
+            & pl.col("vehicle.trip.trip_id").is_not_null()
+            & pl.col("vehicle.vehicle.id").is_not_null()
+            & pl.col("vehicle.timestamp").is_not_null()
+        )
+        .select(
+            pl.col("vehicle.trip.route_id").cast(pl.String).alias("route_id"),
+            pl.col("vehicle.trip.trip_id").cast(pl.String).alias("trip_id"),
+            pl.col("vehicle.stop_id").cast(pl.String).alias("stop_id"),
+            pl.col("vehicle.current_stop_sequence").cast(pl.String).alias("stop_sequence"),
+            pl.col("vehicle.trip.direction_id")
+            .cast(pl.Int8)
+            .alias("direction_id"),
+            pl.col("vehicle.trip.start_time")
+            .cast(pl.String)
+            .alias("start_time"),
+            pl.col("vehicle.trip.start_date")
+            .cast(pl.String)
+            .alias("service_date"),
+            pl.col("vehicle.vehicle.id").cast(pl.String).alias("vehicle_id"),
+            pl.col("vehicle.vehicle.label").cast(pl.String).alias("vehicle_label"),
+            pl.col("vehicle.current_status")
+            .cast(pl.String)
+            .alias("current_status"),
+            pl.from_epoch("vehicle.timestamp").alias("vehicle_timestamp"),
+        )
+        .with_columns(
+            pl.when(pl.col("current_status") == "INCOMING_AT")
+            .then(pl.lit("IN_TRANSIT_TO"))
+            .otherwise(pl.col("current_status"))
+            .cast(pl.String)
+            .alias("current_status"),
+        )
+        .sort(["vehicle_id", "vehicle_timestamp"])
+        .collect()
+    )
+
+    if vehicle_positions.is_empty():
+        # if there are now elements, adjust the columns to match the return
+        # dataframe schema.
+        vehicle_events = vehicle_positions.with_columns(
+            pl.lit(None).cast(pl.Datetime).alias("arrival_gtfs"),
+            pl.lit(None).cast(pl.Datetime).alias("travel_towards_gtfs"),
+        ).drop("current_status", "vehicle_timestamp")
+
+    else:
+        # using the vehicle positions dataframe, create a dataframe for each
+        # event by pivoting and mapping the current status onto arrivals and
+        # departures. initially, they map to arrival times and traveling
+        # towards times. shift the traveling towards datetime over trip id to
+        # get the departure time.
+        vehicle_events = vehicle_positions.pivot(
+            values="vehicle_timestamp",
+            aggregate_function="min",
+            index=[
+                "route_id",
+                "direction_id",
+                "trip_id",
+                "stop_id",
+                "stop_sequence",
+                "start_time",
+                "service_date",
+                "vehicle_id",
+                "vehicle_label"
+            ],
+            columns="current_status",
```

Comment:
Coming out of this `pivot` operations, we're not 100% guaranteed to have a `STOPPED_AT` and `IN_TRANSIT_TO` column, so we should be validating that and or/adding the columns if they don't exist before the `rename` operation.

Reply:
i split the main ingestion function into two functions to make testing easier.

in the pivot method, i check for columns, add null columns if one (or both) is missing, rename, and then run a select to ensure the order is always the same.